File: ArraysAndHashing/Java/ContainsDuplicate/DuplicateHashSetLength.java
/*
 * Time complexity: O(n)
 *   - .distinct() is O(n) (uses a HashSet internally)
 *   - .count() is O(n), but both loops are sequential so total is O(n)
 *
 * Space complexity:
 * - Output: O(1) - only a boolean is returned
 * - Auxiliary: O(n) - a HashSet is used internally by .distinct() to store unique elements
 * - Total: O(n)
 *
 * The logic:
 * We use Java Streams .distinct() to collect unique elements (using a HashSet internally).
 * If the count of distinct elements is less than the length of the input,
 * there is a duplicate.
 */

package Java.ContainsDuplicate;

import java.util.Arrays;

public class DuplicateHashSetLength {
  public static boolean hasDuplicate(int[] nums) {
    return Arrays.stream(nums).distinct().count() < nums.length;
  }

  public static void main(String[] args) {
    int[] arr = { 1, 2, 3, 3 };
    System.out.println(hasDuplicate(arr));

    int[] arr2 = { 1, 2, 3, 4 };
    System.out.println(hasDuplicate(arr2));
  }

}

File: ArraysAndHashing/Python/ContainsDuplicate/duplicate_hash_set_length.py
'''
 * Time complexity: O(n)
 *   - set(nums) is O(n) (builds a hash set)
 *   - len() checks are O(1)
 *
 * Space complexity:
 * - Output: O(1) - only a boolean is returned
 * - Auxiliary: O(n) - a set is created to store unique elements
 * - Total: O(n)
 *
 * The logic:
 * We use Python's set to collect unique elements.
 * If the number of unique elements is less than the length of the input,
 * there is a duplicate.
'''

from typing import List

class Solution:
    def hasDuplicate(self, nums: List[int]) -> bool:
      return len(set(nums)) < len(nums)

if __name__ == "__main__":
    s = Solution()
    print(s.hasDuplicate([1, 2, 9, 15, 15]))
